### PR TITLE
React Page Migration: /courses/:id

### PR DESCRIPTION
**Pull request by syoopie**
Migrate course home page

**Known Issues:**
- Tables are rendered twice on initial load due to useEffect needed for the "see more" function (negligible since it is at most 3 * 5 sets of data

**Announcements:**
image

**Pending Todos:**
image

**Latest Activity:**
image

**Not enrolled in a course:**
image

**Personalized timeline tooltips:**
image

### CODE REVIEW

File: app/views/course/announcements/_announcement.json.jbuilder [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
```
Comment (ekowidianto) [inline app/views/course/announcements/_announcement.json.jbuilder:1]: Rename to announcement_list_data

File: app/views/announcements/_announcement.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
```
Comment (ekowidianto) [inline app/views/announcements/_announcement.json.jbuilder:1]: Rename to announcement_data

File: client/app/types/course/announcements.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { Permissions } from 'types';
+
+export type AnnouncementDataPermissions = Permissions<'canEdit' | 'canDelete'>;
+
+export interface AnnouncementData {
```
Comment (ekowidianto) [inline client/app/types/course/announcements.ts:5]: Not AnnouncementListData?

Not sure why AnnouncementMiniEntity is not defined as well

File: client/app/types/course/notifications.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { InstanceUserListData } from 'types/instanceUsers';
+import { CourseUserListData } from './course_users';
```
Comment (ekowidianto) [inline client/app/types/course/notifications.ts:2]: Help to rename the file to courseUsers

**Review by ekowidianto — CHANGES_REQUESTED** [review]
Just did a very quick review.. I will do another round once you have finalised the PR.
Some comments below and as discussed offline:
- Implement see more before doing backend pagination
- Go through all the existing view logic (eg enter password vs resume)
- Delete all unused files
- Fix backend N+1 problems by preloading the data and store it into a hash
- Frontend performance problem when clicking ignore button - all tables on the page are re-rendered which should not be the case. Only the affected row should be updated

File: app/views/announcements/_announcement_data.json.jbuilder [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# frozen_string_literal: true
```
[the remark this answers is not in the document]
Comment (syoopie) [reply app/views/announcements/_announcement_data.json.jbuilder:1]: The file is not needed. Will remove it first

File: client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    setEnd(end + 5);
+  };
+
+  return (
+    <>
+      <h2>{header}</h2>
+      <Table>
+        <TableHead>
+          <TableRow>
+            <TableCell>Title</TableCell>
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx:164]: React intl.

Go through everything again and make sure everything is intlized

File: client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+      {todos.length > shavedTodos.length && (
+        <div
+          style={{
+            display: 'flex',
+            justifyContent: 'center',
+            padding: 10,
+            marginTop: 10,
+          }}
+        >
+          <Button onClick={handleSeeMore}>SEE MORE</Button>
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx:213]: react intl and include how many more

File: client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+          accessButtonText = 'Attempt';
+          accessButtonLink = getAssessmentSubmissionURL(
+            getCourseId(),
+            todo.actableId,
+          );
+        }
+        break;
+
+      case 'videos':
+        accessButtonText = 'Watch';
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx:95]: intl

File: client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  return (
+    <Button
+      id={`todo-ignore-button-${todoId}`}
+      variant="outlined"
+      color="secondary"
+      onClick={(): void => {
+        onIgnore();
+      }}
+    >
+      Ignore
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx:56]: intl

File: client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  },
+});
+
+const TodoIgnoreButton: FC<Props> = (props) => {
+  const { intl, ignoreLink, todoType, todoId } = props;
+  const dispatch = useDispatch<AppDispatch>();
+
+  const onIgnore = (): Promise<void> => {
+    const courseId = getCourseId();
+    if (courseId) {
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx:33]: Why do we need a check here?

File: client/app/lib/components/StartEndTime.tsx [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+const CustomTooltip = styled(({ className, ...props }: TooltipProps) => (
+  <Tooltip {...props} arrow classes={{ popper: className }} />
+))({
+  [`& .${tooltipClasses.tooltip}`]: {
+    maxWidth: 500,
+  },
+});
+
+/*
+This compoenent is meant for displaying effective and reference time differences
```
Comment (ekowidianto) [inline client/app/lib/components/StartEndTime.tsx:35]: component

File: client/app/bundles/course/courses/components/forms/CourseInvitationCodeForm.tsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    if (code.trim() === '') {
+      toast.error(intl.formatMessage(translations.emptyCodeFailure));
+      return new Promise<void>(() => {});
+    }
+    const registrationLink = getRegistrationURL(getCourseId());
+    const data = new FormData();
+    data.append('registration[code]', code);
+    return dispatch(sendNewRegistrationCode(registrationLink, data))
+      .then(() => {
+        window.location.reload();
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/forms/CourseInvitationCodeForm.tsx:49]: Can i check why do we reload here? I think we can just make a json api call to show index instead and replace the data in the store?

**Review by ekowidianto — CHANGES_REQUESTED** [review]
Good work @syoopie ! Some comments below!

**Comment by ekowidianto** [comment]
@syoopie I think theres some unused files (including scss files) thats not deleted yet (eg _registration.html.slim). Could you take a look?

File: client/app/bundles/course/announcements/components/AnnouncementCard.tsx [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        // toast.success(intl.formatMessage(translations.deletionSuccess));
+        window.location.href = `${getCourseURL(getCourseId())}/announcements`;
+      })
+      .catch((error) => {
+        toast.error(intl.formatMessage(translations.deletionFailure));
+        throw error;
+      });
+  };
+
+  const onEdit = (): void => {
```
[the remark this answers is not in the document]
Comment (syoopie) [reply client/app/bundles/course/announcements/components/AnnouncementCard.tsx:56]: Yup

File: client/app/types/course/courses.ts [hunk 14]
```diff
@@ -1,23 +1,80 @@
...
-
-export type CoursePermissions = Permissions<'canCreate'>;
+import { AnnouncementListData, AnnouncementMiniEntity } from './announcements';
+import { CourseUserListData } from './courseUsers';
+import { TodoData } from './lesson-plan/todos';
+import { NotificationData } from './notifications';
 
 export type CourseListPermissions = Permissions<'canCreate'>;
 
+export type CourseDataPermissions = Permissions<
```
[the remark this answers is not in the document]
Comment (syoopie) [reply client/app/types/course/courses.ts:9]: This is only used in (Data / Entity / Courses Home page) and not (ListData / MiniEntity / All courses) though

[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/app/types/course/courses.ts:9]: Got it!

File: client/app/bundles/course/courses/reducers.ts [hunk 15]
```diff
@@ -21,6 +33,66 @@ const reducer = produce((draft: CoursesState, action: CoursesActionType) => {
...
+      const courseId = action.courseId;
+      const todoType = action.todoType;
+      const todoId = action.todoId;
+      const course = draft.courses.byId[courseId];
+      // Inefficient: filter is O(n)
+      if (course) {
+        switch (todoType) {
+          case 'assessments':
+            if (course.assessmentTodos) {
+              course.assessmentTodos = course.assessmentTodos.filter(
```
[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/app/bundles/course/courses/reducers.ts:53]: I mean when you structure the data in the backend, it should be already in dict format? what do you think?

File: client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx (hunk 10 above)
Comment (syoopie) [reply client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx:33]: courseId can be string | null. Without the check there will be an error
image

File: client/app/bundles/course/courses/components/forms/CourseInvitationCodeForm.tsx (hunk 12 above)
Comment (syoopie) [reply client/app/bundles/course/courses/components/forms/CourseInvitationCodeForm.tsx:49]: Need to re-render the entire page for the rails stuff to render (breadcrum, sidebar)

Comment (ekowidianto) [reply client/app/bundles/course/courses/components/forms/CourseInvitationCodeForm.tsx:49]: got it thanks!

File: client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx (hunk 10 above)
Comment (ekowidianto) [reply client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx:33]: Ah i see, I think I am quite sure we courseId will never be null right?  U can do `getCourseId()!`

Comment (syoopie) [reply client/app/bundles/course/courses/components/buttons/TodoIgnoreButton.tsx:33]: It is already using getCourseId() haha but the function is declared in a JS file and the type is returned as string | null

**Review by ekowidianto — APPROVED** [review]
Looks good thanks! Can re-squash the PR?

File: client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+        if (todo.progress === 'not_started') {
+          if (!todo.canAccess && todo.canAttempt) {
+            accessButtonText = intl.formatMessage(
+              translations.accessButtonEnterPassword,
+            );
+            accessButtonLink = getAssessmentURL(
+              getCourseId(),
+              todo.itemActableId,
+            );
+          } else {
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx:125]: not elsif todo.canAttempt ?

image

seems like it should be a post request, not get.

File: client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+            accessButtonText = intl.formatMessage(
+              translations.accessButtonAttempt,
+            );
+            accessButtonLink = getAssessmentSubmissionURL(
+              getCourseId(),
+              todo.itemActableId,
+            );
+          }
+        } else {
```
Comment (ekowidianto) [inline client/app/bundles/course/courses/components/tables/PendingTodosTable.tsx:134]: elsif in progress?

**Pull request merged**